xapian-core/cluster/kmeans.cc

```diff
@@ -24,120 +24,126 @@
+}
+
 string
 KMeans::get_description() const {
     LOGCALL(API, string, "KMeans::get_description()", NO_ARGS);
     return "KMeans clusterer";
 }
 
 void
-KMeans::initialize_centroids(ClusterSet &cset) {
-    LOGCALL_VOID(API, "KMeans::initialize_centroids()", cset);
-    set<docid> centroid_docs;
+KMeans::initialize_clusters(ClusterSet &cset) {
+    LOGCALL_VOID(API, "KMeans::initialize_clusters()", cset);
     int size = docs.size();
-    srand(time(NULL));
+    mt19937 generator;
+    generator.seed(random_device()());
+    uniform_int_distribution<mt19937::result_type> random(0, size);
     for (unsigned int i = 0; i < k;) {
-	int x = rand() % size;
-	if (centroid_docs.find(x) == centroid_docs.end()) {
-	    Point a = docs[x];
-	    Centroid centroid;
-	    centroid.set_to_point(a);
-	    centroids.push_back(centroid);
-	    centroid_docs.insert(x);
-	    Cluster cluster1(centroid);
-	    cset.add_cluster(cluster1);
-	    ++i;
-	}
+	int x = random(generator);
+	Centroid centroid(docs[x]);
+	Cluster cluster(centroid);
+	cset.add_cluster(cluster);
+	++i;
     }
 }
 
 void
-KMeans::initialize_points(const MSet &source, TermListGroup &tlg) {
-    LOGCALL_VOID(API, "KMeans::initialize_points()", source | tlg);
+KMeans::initialize_points(const MSet &source) {
+    LOGCALL_VOID(API, "KMeans::initialize_points()", source);
+    TermListGroup tlg(source);
     for (MSetIterator it = source.begin(); it != source.end(); ++it) {
 	Point p;
 	p.initialize(tlg, it.get_document());
 	docs.push_back(p);
     }
 }
 
-bool
-KMeans::converge(vector<Centroid> &previous, vector<Centroid> &current) const {
-    LOGCALL(API, bool, "KMeans::converge()", previous | current);
-    CosineDistance cosine;
-    for (unsigned int i = 0; i < k; ++i) {
-	Centroid &prev = previous[i];
-	Centroid &curr = current[i];
-	double dist = cosine.similarity(prev, curr);
-	if (dist > EPSILON)
-	    return false;
-    }
-    return true;
-}
-
 ClusterSet
 KMeans::cluster(MSet &mset) {
     LOGCALL(API, ClusterSet, "KMeans::cluster()", mset);
     unsigned int size = mset.size();
+    if (k >= size)
+	k = size;
     mset.fetch();
-    if (size <= 0)
-	throw AssertionError("Size of MSet should be greater than zero");
-    if (k <= 0)
-	throw AssertionError("Number of clusters should be greater than zero");
-    if (k > size)
-	throw AssertionError("The number of clusters cannot be greater than number of documents in MSet");
+    initialize_points(mset);
     ClusterSet cset;
-    TermListGroup tlg;
-    tlg.add_documents(mset);
-    CosineDistance cosine;
-    initialize_points(mset, tlg);
-    initialize_centroids(cset);
-    double min = INT_MAX;
-    double min_cluster = 0;
-
-    vector<Centroid> prev;
-    for (int i = 0; i < max_iters; ++i) {
+    initialize_clusters(cset);
+    for (unsigned int i = 0; i < max_iters; ++i) {
+	/** Assign each point to ihe cluster corresponding to its
+	 *  closest cluster centroid
+	 */
+	cset.clear_clusters();
+	CosineDistance distance;
```

Comment:
This is now created in every loop. This isn't actually a huge cost, but it seems unnecessary; where you had it before was fine, wasn't it?